packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -3707,6 +3705,25 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
     _textLayoutLastMaxWidth = maxWidth;
   }
 
+  // Computes the text metrics if `_textPainter`'s layout information was marked
+  // as dirty.
+  //
+  // This method must be called in [RenderEditable]'s public methods that
+  // expose `_textPainter`'s metrics. For instance, `systemFontsDidChange` sets
+  // _textPainter._paragraph to null, accessing _textPainter's metrics
+  // immediately after `systemFontsDidChange` without first calling this method
+  // may crash.
+  //
+  // This method is also called in various paint methods (`RenderEditable.paint`
+  // as well as its foreground/background painters' `paint`). It's needed
+  // because invisible render objects kept in the tree by `KeepAlive` may not
+  // get a chance to do layout but can still paint.
+  // See https://github.com/flutter/flutter/issues/84896.
+  void _computeTextMetricsIfNeeded() {
```

Comment:
Why is it called "IfNeeded"?  I expected some kind of conditional in this method.

Reply:
The condition lives in `TextPainter.layout`: https://github.com/flutter/flutter/blob/a1a360bf1385d82ef99319fed06352a2d0189b21/packages/flutter/lib/src/painting/text_painter.dart#L582-L583. I added a few more lines to the code comments.